sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ClassAndMethodName.cs

```diff
@@ -0,0 +1,339 @@
+
+        private static readonly DiagnosticDescriptor rule_MethodName =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId_TypeName, MessageFormat, RspecStrings.ResourceManager);
+
+        private static readonly DiagnosticDescriptor rule_TypeName =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId_MethodName, MessageFormat, RspecStrings.ResourceManager);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+            => ImmutableArray.Create(rule_MethodName, rule_TypeName);
+
+        private static readonly Dictionary<SyntaxKind, string> TypeKindNameMapping = new Dictionary<SyntaxKind, string>
+        {
+            {SyntaxKind.StructDeclaration, "struct" },
+            {SyntaxKind.ClassDeclaration, "class" },
+            {SyntaxKind.InterfaceDeclaration, "interface" },
+            {SyntaxKind.MethodDeclaration, "method" },
+            {SyntaxKind.PropertyDeclaration, "property" }
+        };
+
+        protected sealed override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                CheckTypeName,
+                SyntaxKind.ClassDeclaration,
+                SyntaxKind.InterfaceDeclaration,
+                SyntaxKind.StructDeclaration);
+
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var declaration = (MethodDeclarationSyntax)c.Node;
+                    CheckMemberName(declaration, declaration.Identifier, c);
+                },
+                SyntaxKind.MethodDeclaration);
+
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var declaration = (PropertyDeclarationSyntax)c.Node;
+                    CheckMemberName(declaration, declaration.Identifier, c);
+                },
+                SyntaxKind.PropertyDeclaration);
+        }
+
+        private static void CheckTypeName(SyntaxNodeAnalysisContext context)
+        {
+            var typeDeclaration = (BaseTypeDeclarationSyntax)context.Node;
+            var identifier = typeDeclaration.Identifier;
+            var symbol = context.SemanticModel.GetDeclaredSymbol(typeDeclaration);
+
+            if (IsTypeComRelated(symbol))
+            {
+                return;
+            }
+
+            if (identifier.ValueText.StartsWith("_", StringComparison.Ordinal) ||
+                identifier.ValueText.EndsWith("_", StringComparison.Ordinal))
+            {
+                context.ReportDiagnostic(Diagnostic.Create(rule_TypeName, identifier.GetLocation(), TypeKindNameMapping[typeDeclaration.Kind()],
+                    identifier.ValueText, MessageFormatUnderscore));
+                return;
+            }
+
+            string suggestion;
+            bool isNameValid = IsTypeNameValid(identifier.ValueText,
+                typeDeclaration is InterfaceDeclarationSyntax,
```

Comment:
I would suggest to prepend the parameter name to the variable .
`areUnderscoresAllowed: context.IsTest()`.